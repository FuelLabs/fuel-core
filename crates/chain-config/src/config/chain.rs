use fuel_core_storage::MerkleRoot;
use fuel_core_types::{
    blockchain::header::StateTransitionBytecodeVersion,
    fuel_crypto::Hasher,
    fuel_tx::ConsensusParameters,
    fuel_types::{
        fmt_truncated_hex,
        AssetId,
    },
};
use serde::{
    Deserialize,
    Serialize,
};
#[cfg(feature = "std")]
use std::fs::File;
#[cfg(feature = "std")]
use std::path::Path;

use crate::{
    genesis::GenesisCommitment,
    ConsensusConfig,
};

#[cfg(feature = "std")]
use crate::SnapshotMetadata;

pub const LOCAL_TESTNET: &str = "local_testnet";
pub const BYTECODE_NAME: &str = "state_transition_bytecode.wasm";

#[derive(Clone, derivative::Derivative, Deserialize, Serialize, Eq, PartialEq)]
#[derivative(Debug)]
pub struct ChainConfig {
    pub chain_name: String,
    pub consensus_parameters: ConsensusParameters,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    pub genesis_state_transition_version: Option<StateTransitionBytecodeVersion>,
    /// Note: The state transition bytecode is stored in a separate file
    /// under the `BYTECODE_NAME` name in serialization form.
    #[serde(skip)]
    #[derivative(Debug(format_with = "fmt_truncated_hex::<16>"))]
    pub state_transition_bytecode: Vec<u8>,
    pub consensus: ConsensusConfig,
}

#[cfg(feature = "test-helpers")]
impl Default for ChainConfig {
    fn default() -> Self {
        Self {
            chain_name: "local".into(),
            consensus_parameters: ConsensusParameters::default(),
            genesis_state_transition_version: Some(
                fuel_core_types::blockchain::header::LATEST_STATE_TRANSITION_VERSION,
            ),
            // Note: It is invalid bytecode.
            state_transition_bytecode: vec![123; 1024],
            consensus: ConsensusConfig::default_poa(),
        }
    }
}

impl ChainConfig {
    pub const BASE_ASSET: AssetId = AssetId::zeroed();

    #[cfg(feature = "std")]
    pub fn load(path: impl AsRef<Path>) -> anyhow::Result<Self> {
        use std::io::Read;
        let path = path.as_ref();
        let mut json = String::new();
        std::fs::File::open(path)?.read_to_string(&mut json)?;
        let mut chain_config: ChainConfig =
            serde_json::from_str(json.as_str()).map_err(|e| {
                anyhow::Error::new(e).context(format!(
                    "an error occurred while loading the chain state file: {:?}",
                    path.to_str()
                ))
            })?;

        let bytecode_path = path.with_file_name(BYTECODE_NAME);
        let bytecode = std::fs::read(bytecode_path).map_err(|e| {
            anyhow::Error::new(e).context(format!(
                "an error occurred while loading the state transition bytecode: {:?}",
                path.to_str()
            ))
        })?;

        chain_config.state_transition_bytecode = bytecode;

        Ok(chain_config)
    }

    #[cfg(feature = "std")]
    pub fn from_snapshot_metadata(
        snapshot_metadata: &SnapshotMetadata,
    ) -> anyhow::Result<Self> {
        Self::load(&snapshot_metadata.chain_config)
    }

    #[cfg(feature = "std")]
    pub fn write(&self, path: impl AsRef<Path>) -> anyhow::Result<()> {
        use anyhow::Context;

        let bytecode_path = path.as_ref().with_file_name(BYTECODE_NAME);
        let chain_config_file = File::create(path)?;

        serde_json::to_writer_pretty(chain_config_file, self)
            .context("failed to dump chain parameters snapshot to JSON")?;
        std::fs::write(bytecode_path, &self.state_transition_bytecode)
            .context("failed to write state transition bytecode")?;

        Ok(())
    }

    #[cfg(feature = "test-helpers")]
    pub fn local_testnet() -> Self {
        Self {
            chain_name: LOCAL_TESTNET.to_string(),
            ..Default::default()
        }
    }
}

impl GenesisCommitment for ChainConfig {
    fn root(&self) -> anyhow::Result<MerkleRoot> {
        let chain_config_bytes =
            postcard::to_allocvec(&self).map_err(anyhow::Error::msg)?;
        let config_hash = *Hasher::default()
            .chain(chain_config_bytes.as_slice())
            .chain(self.state_transition_bytecode.as_slice())
            .finalize();

        Ok(config_hash)
    }
}

#[cfg(test)]
mod tests {
    #[cfg(feature = "std")]
    use std::env::temp_dir;

    use super::ChainConfig;

    #[cfg(feature = "std")]
    #[test]
    fn can_roundtrip_write_and_read() {
        let tmp_dir = temp_dir();
        let file = tmp_dir.join("config.json");

        let disk_config = ChainConfig::local_testnet();
        disk_config.write(&file).unwrap();

        let load_config = ChainConfig::load(&file).unwrap();

        assert_eq!(disk_config, load_config);
    }

    #[test]
    fn snapshot_local_testnet_config() {
        let config = ChainConfig::local_testnet();
        let json = serde_json::to_string_pretty(&config).unwrap();
        insta::assert_snapshot!(json);
    }
}
