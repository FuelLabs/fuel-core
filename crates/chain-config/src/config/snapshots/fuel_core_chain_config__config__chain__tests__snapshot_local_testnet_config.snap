---
source: crates/chain-config/src/config/chain.rs
expression: json
---
{
  "chain_name": "local_testnet",
  "consensus_parameters": {
    "V1": {
      "tx_params": {
        "V1": {
          "max_inputs": 255,
          "max_outputs": 255,
          "max_witnesses": 255,
          "max_gas_per_tx": 100000000,
          "max_size": 112640,
          "max_bytecode_subsections": 255
        }
      },
      "predicate_params": {
        "V1": {
          "max_predicate_length": 1048576,
          "max_predicate_data_length": 1048576,
          "max_message_data_length": 1048576,
          "max_gas_per_predicate": 100000000
        }
      },
      "script_params": {
        "V1": {
          "max_script_length": 1048576,
          "max_script_data_length": 1048576
        }
      },
      "contract_params": {
        "V1": {
          "contract_max_size": 102400,
          "max_storage_slots": 255
        }
      },
      "fee_params": {
        "V1": {
          "gas_price_factor": 1000000000,
          "gas_per_byte": 4
        }
      },
      "chain_id": 0,
      "gas_costs": {
        "V2": {
          "add": 1,
          "addi": 1,
          "and": 1,
          "andi": 1,
          "bal": 13,
          "bhei": 1,
          "bhsh": 1,
          "burn": 132,
          "cb": 1,
          "cfei": 1,
          "cfsi": 1,
          "div": 1,
          "divi": 1,
          "eck1": 951,
          "ecr1": 3000,
          "ed19": 3000,
          "eq": 1,
          "exp": 1,
          "expi": 1,
          "flag": 1,
          "gm": 1,
          "gt": 1,
          "gtf": 1,
          "ji": 1,
          "jmp": 1,
          "jne": 1,
          "jnei": 1,
          "jnzi": 1,
          "jmpf": 1,
          "jmpb": 1,
          "jnzf": 1,
          "jnzb": 1,
          "jnef": 1,
          "jneb": 1,
          "lb": 1,
          "log": 9,
          "lt": 1,
          "lw": 1,
          "mint": 135,
          "mlog": 1,
          "mod": 1,
          "modi": 1,
          "move": 1,
          "movi": 1,
          "mroo": 2,
          "mul": 1,
          "muli": 1,
          "mldv": 1,
          "noop": 1,
          "not": 1,
          "or": 1,
          "ori": 1,
          "poph": 2,
          "popl": 2,
          "pshh": 2,
          "pshl": 2,
          "ret_contract": 13,
          "rvrt_contract": 13,
          "sb": 1,
          "sll": 1,
          "slli": 1,
          "srl": 1,
          "srli": 1,
          "srw": 12,
          "sub": 1,
          "subi": 1,
          "sw": 1,
          "sww": 67,
          "time": 1,
          "tr": 105,
          "tro": 60,
          "wdcm": 1,
          "wqcm": 1,
          "wdop": 1,
          "wqop": 1,
          "wdml": 1,
          "wqml": 1,
          "wddv": 1,
          "wqdv": 2,
          "wdmd": 3,
          "wqmd": 4,
          "wdam": 2,
          "wqam": 3,
          "wdmm": 3,
          "wqmm": 3,
          "xor": 1,
          "xori": 1,
          "aloc": {
            "LightOperation": {
              "base": 2,
              "units_per_gas": 214
            }
          },
          "call": {
            "LightOperation": {
              "base": 144,
              "units_per_gas": 214
            }
          },
          "ccp": {
            "LightOperation": {
              "base": 15,
              "units_per_gas": 103
            }
          },
          "croo": {
            "LightOperation": {
              "base": 1,
              "units_per_gas": 1
            }
          },
          "csiz": {
            "LightOperation": {
              "base": 17,
              "units_per_gas": 790
            }
          },
          "k256": {
            "LightOperation": {
              "base": 11,
              "units_per_gas": 214
            }
          },
          "ldc": {
            "LightOperation": {
              "base": 15,
              "units_per_gas": 272
            }
          },
          "logd": {
            "LightOperation": {
              "base": 26,
              "units_per_gas": 64
            }
          },
          "mcl": {
            "LightOperation": {
              "base": 1,
              "units_per_gas": 3333
            }
          },
          "mcli": {
            "LightOperation": {
              "base": 1,
              "units_per_gas": 3333
            }
          },
          "mcp": {
            "LightOperation": {
              "base": 1,
              "units_per_gas": 2000
            }
          },
          "mcpi": {
            "LightOperation": {
              "base": 3,
              "units_per_gas": 2000
            }
          },
          "meq": {
            "LightOperation": {
              "base": 1,
              "units_per_gas": 2500
            }
          },
          "retd_contract": {
            "LightOperation": {
              "base": 29,
              "units_per_gas": 62
            }
          },
          "s256": {
            "LightOperation": {
              "base": 2,
              "units_per_gas": 214
            }
          },
          "scwq": {
            "LightOperation": {
              "base": 13,
              "units_per_gas": 5
            }
          },
          "smo": {
            "LightOperation": {
              "base": 209,
              "units_per_gas": 55
            }
          },
          "srwq": {
            "LightOperation": {
              "base": 47,
              "units_per_gas": 5
            }
          },
          "swwq": {
            "LightOperation": {
              "base": 44,
              "units_per_gas": 5
            }
          },
          "contract_root": {
            "LightOperation": {
              "base": 75,
              "units_per_gas": 1
            }
          },
          "state_root": {
            "LightOperation": {
              "base": 412,
              "units_per_gas": 1
            }
          },
          "new_storage_per_byte": 1,
          "vm_initialization": {
            "HeavyOperation": {
              "base": 2000,
              "gas_per_unit": 0
            }
          }
        }
      },
      "base_asset_id": "0000000000000000000000000000000000000000000000000000000000000000",
      "block_gas_limit": 100000000,
      "privileged_address": "0000000000000000000000000000000000000000000000000000000000000000"
    }
  },
  "genesis_state_transition_version": 3,
  "consensus": {
    "PoA": {
      "signing_key": "22ec92c3105c942a6640bdc4e4907286ec4728e8cfc0d8ac59aad4d8e1ccaefb"
    }
  }
}
